[17.0][MIG] mail_tracking_mailgun: Migration to 17.0

chienandalu opened the pull request:
Pending:

- [x] <del>Doing practical tests with sandbox API keys</del>

cc @Tecnativa TT49914

antespi committed:
[ADD] mail_tracking_mailgun

antespi committed:
[8.0][IMP][mail_tracking] Speed installation time and discard concurrent events (#82)

[IMP] mail_tracking: Speed installation time, discard concurrent events and other fixes

antespi committed:
[MIG] mail_tracking_mailgun

antespi committed:
OCA Transbot updated translations from Transifex

pedrobaeza committed:
[MIG] Make modules uninstallable

pedrobaeza committed:
[MIG] Rename manifest files

bouvyd committed:
[MIG] mail_tracking_mailgun: Migrated to 10.0

oca-transbot committed:
OCA Transbot updated translations from Transifex

rafaelbn committed:
Add Contributors

Both Carlos and me have work a lot in this module, maybe not coding but much testing and definition in usability

oca-transbot committed:
OCA Transbot updated translations from Transifex

chienandalu committed:
[IMP] mute logger on tests error responses so Travis doesn't fail

oca-transbot committed:
OCA Transbot updated translations from Transifex

chienandalu committed:
[10.0][IMP] mail_tracking_mailgun: add partner mail checks

oca-transbot committed:
OCA Transbot updated translations from Transifex

chienandalu committed:
[10.0][FIX] mail_tracking_mailgun: Avoid key dict errors

yajo committed:
[FIX] mail_tracking_mailgun: Avoid errors without event

When this method is called without an event (a.k.a. `event=None`), this method produces the following error:

```
Traceback (most recent call last):
  File "/opt/odoo/auto/addons/mail/models/mail_mail.py", line 278, in send
    res = IrMailServer.send_email(msg, mail_server_id=mail.mail_server_id.id)
  File "/opt/odoo/auto/addons/mail_tracking/models/ir_mail_server.py", line 88, in send_email
    tracking_email.smtp_error(self, smtp_server_used, e)
  File "/opt/odoo/auto/addons/mail_tracking/models/mail_tracking_email.py", line 213, in smtp_error
    self.sudo()._partners_email_bounced_set('error')
  File "/opt/odoo/auto/addons/mail_tracking/models/mail_tracking_email.py", line 203, in
_partners_email_bounced_set
    ]).email_bounced_set(self, reason, event=event)
  File "/opt/odoo/auto/addons/mail_tracking_mailgun/models/res_partner.py", line 22, in
email_bounced_set
    self._email_bounced_set(reason, event)
  File "/opt/odoo/auto/addons/mail_tracking_mailgun/models/res_partner.py", line 33, in
_email_bounced_set
    event['Message-Id'] or '') TypeError: 'NoneType' object has no attribute '__getitem__'
```

So, we now assume we do not always have an event.

oca-transbot committed:
OCA Transbot updated translations from Transifex

pedrobaeza committed:
[FIX] mail_tracking_mailgun: Message error on bounce

chienandalu committed:
[FIX] mail_tracking_mailgun: manual sync gets events from other recipients

chienandalu committed:
[IMP] mail_tracking_mailgun: validation auto check

- Configurable partner email auto check.

oca-transbot committed:
OCA Transbot updated translations from Transifex

chienandalu committed:
[MIG] mail_tracking_mailgun: Migration to 11.0

sergio-teruel committed:
[11.0][FIX] mail_tracking_mailgun: Read config parameters with sudo

didierdonze committed:
Translated using Weblate (French)

Currently translated at 16.7% (4 of 24 strings)

Translation: social-11.0/social-11.0-mail_tracking_mailgun
Translate-URL: https://translation.odoo-community.org/projects/social-11-0/social-11-0-mail_tracking_mailgun/fr/

chienandalu committed:
[FIX] mail_tracking_mailgun: alternative domain

- In case the sending domain is different from the one configured in the
mail.domain.catchall setting.

ernestotejeda committed:
[MIG] mail_tracking_mailgun: Migration to 12.0

zaoral committed:
[ADD] mail_tracking_mailgun: manage failed state from mailgun

This change let odoo process a state sent from mailgun legacy webhooks
that seems to apply when the message is not sent because the related
email has been mark us as spam or have bounced before. For solve this
add two new states to _mailgun_event_type_mapping_mailgun_event_type_mapping
method:

* failed: Mailgun could not deliver the email to the recipient email server
* rejected: Mailgun rejected the request to send/forward the email

source in https://documentation.mailgun.com/en/latest/quickstart-events.html#events.

chienandalu committed:
[IMP] mail_tracking_mailgun: link to bounce event

pedrocs-exo committed:
Translated using Weblate (Portuguese)

Currently translated at 95.8% (23 of 24 strings)

Translation: social-12.0/social-12.0-mail_tracking_mailgun
Translate-URL: https://translation.odoo-community.org/projects/social-12-0/social-12-0-mail_tracking_mailgun/pt/

sysadminmatmoz committed:
Translated using Weblate (Slovenian)

Currently translated at 12.5% (3 of 24 strings)

Translation: social-12.0/social-12.0-mail_tracking_mailgun
Translate-URL: https://translation.odoo-community.org/projects/social-12-0/social-12-0-mail_tracking_mailgun/sl/

HviorForgeFlow committed:
[IMP] mail_tracking_mailgun: black, isort

HviorForgeFlow committed:
[MIG] mail_tracking_mailgun: Migration to 13.0

hbrunn committed:
[FIX] import mock the python3 way

Commit:
[IMP] pre-commit run -a

Tardo committed:
[IMP] Apply pre-commit changes: Resolve conflicts

CarlosRoca13 committed:
[MIG] mail_tracking_mailgun: Migration to v14.0

Commit:
[IMP] mail_tracking_mailgun: refactor to support modern webhooks

Before this patch, the module was designed after the [deprecated Mailgun webhooks][3]. However Mailgun had the [events API][2] which was quite different. Modern Mailgun has deprecated those webhooks and instead uses new ones that include the same payload as the events API, so you can reuse code.

However, this was incorrectly reusing the code inversely: trying to process the events API through the same code prepared for the deprecated webhooks.

Besides, both `failed` and `rejected` mailgun events were mapped to `error` state, but that was also wrong because [`mail_tracking` doesn't have an `error` state][1].

So the logic of the whole module is changed, adapting it to process the events API payload, both through controllers (prepared for the new webhooks) and manual updates that directly call the events API.

Also, `rejected` is now translated into `reject`, and `failed` is translated into `hard_bounce` or `soft_bounce` depending on the severity, as specified by [mailgun docs][2]. Also, `bounced` and `dropped` mailgun states are removed because they don't exist, and instead `failed` and `rejected` properly get their metadata.

Of course, to know the severity, now the method to obtain that info must change, it' can't be a simple dict anymore.

Added more parameters because for example modern Mailgun uses different keys for signing payload than for accessing the API. As there are so many parameters, configuration is now possible through `res.config.settings`. Go there to autoregister webhooks too.

Since the new webhooks are completely incompatible with the old supposedly-abstract webhooks controllers (that were never really that abstract), support for old webhooks is removed, and it will be removed in the future from `mail_tracking` directly. There is a migration script that attempts to unregister old webhooks and register new ones automatically.

[1]: https://github.com/OCA/social/blob/f73de421e28a43d018176f61725a3a59665f715d/mail_tracking/models/mail_tracking_event.py#L31-L42
[2]: https://documentation.mailgun.com/en/latest/api-events.html#event-types
[3]: https://documentation.mailgun.com/en/latest/api-webhooks-deprecated.html

nicomacr committed:
[MIG] mail_tracking_mailgun: Migration to 15.0

chienandalu committed:
[IMP] mail_tracking_mailgun: mass mailing manual tracking

Mass mailing are tracked from mail.trace as the don't store a message in
the db. In order to gather the message_id and be able to do manual
checks to mailgun, that's the table where we should get the message id.

TT40816

alan196 committed:
[MIG] mail_tracking_mailgun: Migration to 16.0

chienandalu committed:
[FIX] mail_tracking_mailgun: bounces error

Since https://github.com/odoo/odoo/commit/9a47101964a56fe5a7a44da37209e8f67b508395
bouncing the emails through the mailgun controller causes a singleton
error when the anonymous user tries to record the bounce message.

TT44585

Ivorra78 committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (60 of 60 strings)

Translation: social-15.0/social-15.0-mail_tracking_mailgun
Translate-URL: https://translation.odoo-community.org/projects/social-15-0/social-15-0-mail_tracking_mailgun/es/

chienandalu committed:
[FIX/IMP] mail_tracking_mailgun: finish migration

- Fix tests
- Fix warnings
- Improve code
- Get rid of superfluous stuff
- Remove auto-validation. It couldn't be working as it was and it would
  drag performance on contacts creation/write if active.

TT44207

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (58 of 58 strings)

Translation: social-16.0/social-16.0-mail_tracking_mailgun
Translate-URL: https://translation.odoo-community.org/projects/social-16-0/social-16-0-mail_tracking_mailgun/it/

vincent-hatakeyama committed:
Translated using Weblate (French)

Currently translated at 37.9% (22 of 58 strings)

Translation: social-16.0/social-16.0-mail_tracking_mailgun
Translate-URL: https://translation.odoo-community.org/projects/social-16-0/social-16-0-mail_tracking_mailgun/fr/

chienandalu committed:
[IMP] mail_tracking_mailgun: black, isort, prettier

chienandalu committed:
[MIG] mail_tracking_mailgun: Migration to 17.0

TT49914

pedrobaeza commented:
/ocabot migration mail_tracking_mailgun

pedrobaeza commented on mail_tracking_mailgun/models/mail_tracking_email.py:
```diff
@@ -78,8 +78,8 @@ def _mailgun_values(self):
         api_url = icp.get_param("mailgun.api_url", "https://api.mailgun.net/v3")
-        catchall_domain = icp.get_param("mail.catchall.domain")
-        domain = icp.get_param("mailgun.domain", catchall_domain)
+        catchall_domain = self.env["mail.alias.domain"].search([], limit=1).name
```
Shouldn't this be adapted for all the possible domains?

chienandalu replied:
Well, this is done just to get a default value if an specific one isn't set in the settings.

pedrobaeza commented on mail_tracking_mailgun/models/res_partner.py:
```diff
@@ -34,9 +34,13 @@ def _email_bounced_set(self, reason, event):
+                    "Email has been bounced: %(email)s\nReason: "
+                    "%(reason)s\nEvent: %(event_str)s"
+                ),
+                ({"email": partner.email, "reason": reason, "event_str": event_str}),
```
This doesn't seem correctly put (with the weird parenthesis), and `_(` accepts already the parameters as new arguments.

chienandalu replied:
Changed

chienandalu committed:
[FIX] mail_tracking_mailgun: discard non Odoo events

When we use the same Mailgun domain for other services, the email events
for those services we'll be pushed to the Odoo controller as well. We
want to discard them as they're useless to us.

Aside from that, in the case a wrong db is called to the controller, we
better logging the failed request is going to be more useful than
raising an error.

pedrobaeza reviewed (approved):
Thank you!

/ocabot merge nobump

pedrobaeza commented:
/ocabot merge nobump

pedrobaeza commented:
/ocabot merge nobump

pedrobaeza commented:
/ocabot merge nobump
